src/java/htsjdk/samtools/SAMTextWriter.java

```diff
@@ -45,13 +45,76 @@
     private final TextTagCodec tagCodec = new TextTagCodec();
     private final SAMTagUtil tagUtil = new SAMTagUtil();
 
+    private final FlagFieldsOutput flagFieldsOutput;
+
+    /**
+     * This determines how flag fields are represented in the output.
+     *  
+     * In a string FLAG, each character represents one bit with
+     * p=0x1 (paired), P=0x2 (properly paired), u=0x4 (unmapped),
+     * U=0x8 (mate unmapped), r=0x10 (reverse), R=0x20 (mate reverse)
+     * 1=0x40 (first), 2=0x80 (second), s=0x100 (not primary), 
+     * x=0x200 (failure), d=0x400 (duplicate), and S=0x800 (secondary).
+     * This was inspired by 'samtools view -X'.
+     * 
+     * We also output a character when the following bits *are not* set:
+     * m=0x4 (mapped), M=0x8 (mate mapped), f=0x10 (forward), F=0x20 
+     * (mate forward).
+     */
+    public enum FlagFieldsOutput {
+        /**
+         * NONE is here so that when the user specifies a value on the command line (Defaults.FLAG_FIELDS_OUTPUT), 
+         * we use the Defaults.FLAG_FIELDS_OUTPUT value rather than FlagFieldsOutput.DEFAULT.  See the SAMFileWriterFactory.
+         */
+        NONE {
+            @Override
+            public String format(final SAMRecord record) {
+                throw new SAMException("NONE not allowed for the FlagFieldsOutput when writing the SAM flag field.");
+            }
+        },
+        DEFAULT {
+            @Override
+            public String format(final SAMRecord record) {
+                return Integer.toString(record.getFlags());
+            }
+        },
+        HEXADECIMAL {
+            @Override
+            public String format(final SAMRecord record) {
+                return String.format("0x%x", record.getFlags());
+            }
+        },
+        STRING {
+            private final String flag2CharTable = "pPuUrR12sxdS\0\0\0\0"; // when the bit is set
+            private final String notFlag2CharTable = "\0\0mMfF\0\0\0\0\0\0\0\0\0\0"; // when the bit is not set
+
+            private String getStringFlags(final int flag) {
+                String s = "";
+                for (int i = 0; i < 16; ++i) {
+                    if ((flag & 1 << i) != 0) { // the bit is set
+                        if ('\0' != flag2CharTable.charAt(i)) s += flag2CharTable.charAt(i);
+                    }
+                    else if ('\0' != notFlag2CharTable.charAt(i)) { // the bit is not set
+                        if ('\0' != notFlag2CharTable.charAt(i)) s += notFlag2CharTable.charAt(i);
```

Comment:
This needlessly performs the same `'\0' != notFlag2CharTable.charAt(i)` test twice.